### PR TITLE
Unset CUBLAS_WORKSPACE_CONFIG for non-deterministic training and inference

### LINKED ISSUE #19127: CUBLAS_WORKSPACE_CONFIG makes torch slow down
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Other

### Bug

https://github.com/ultralytics/ultralytics/issues/16457 reports the same issue and it was fixed in v8.2.102 https://github.com/ultralytics/ultralytics/pull/16466 by removing line: https://github.com/ultralytics/ultralytics/blob/35f0fde06969ab599379b8f1f5af657be2828dc0/ultralytics/utils/__init__.py#L131
However, since v8.3.2 https://github.com/ultralytics/ultralytics/pull/16583 , the line is added back. It will slow down torch inference speed.

### Environment

```
Ultralytics 8.3.72 🚀 Python-3.10.16 torch-2.5.1 CUDA:0 (NVIDIA A100-SXM4-80GB, 81229MiB)

numpy               ✅ 1.26.4<=2.1.1,>=1.23.0
matplotlib          ✅ 3.10.0>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.5.1>=1.8.0
torch               ✅ 2.5.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.20.1>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 6.1.1
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0
```

### Minimal Reproducible Example

```python
import os
import torch
from ultralytics import YOLO
from diffusers import StableDiffusionXLPipeline

pipe = StableDiffusionXLPipeline.from_pretrained(""stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16).to("cuda")
with torch.inference_mode():
    pipe("test") # warmup
    pipe("test")
    os.environ.pop('CUBLAS_WORKSPACE_CONFIG', None)
    pipe("test")
    os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
    pipe("test")
```

```
100%|█████████████████████████████████████████████████████████| 50/50 [00:06<00:00,  8.26it/s] # warmup
100%|█████████████████████████████████████████████████████████| 50/50 [00:05<00:00,  8.64it/s] # w/ CUBLAS_WORKSPACE_CONFIG
100%|█████████████████████████████████████████████████████████| 50/50 [00:04<00:00, 10.02it/s] # w/o CUBLAS_WORKSPACE_CONFIG
100%|█████████████████████████████████████████████████████████| 50/50 [00:05<00:00,  8.57it/s] # w/ CUBLAS_WORKSPACE_CONFIG
```

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes https://github.com/ultralytics/ultralytics/issues/19127

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary  
Improved management of deterministic settings in training for enhanced flexibility and reduced warnings.  

### 📊 Key Changes  
- Added `unset_deterministic()` method to clean up deterministic training configurations.  
- Modified the model initialization to set `CUBLAS_WORKSPACE_CONFIG` for consistency and avoid warnings.  
- Updated training flow to unset deterministic settings after training completes.  
- Updated documentation to reference the newly added `unset_deterministic()` utility.  

### 🎯 Purpose & Impact  
- 🛠 **Flexibility**: Allows better control of deterministic training settings, making it easier to switch configurations dynamically.  
- ⚡ **Reduced Warnings**: Suppresses unnecessary CUDA warnings related to deterministic training, leading to cleaner logs.  
- 🔄 **Streamlined Code**: Simplifies reset of configurations post-training, ensuring an efficient and error-free model training process.  
- 📚 **Clarity**: Documentation updates make the new functionality more accessible to users.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description and link to any relevant [issues](https://github.com/ultralytics/ultralytics/issues). Your reference to https://github.com/ultralytics/ultralytics/issues/19127 is great! Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) if necessary, especially as your changes impact inference performance and environment management.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing. 
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your fix. For example, the removal of `CUBLAS_WORKSPACE_CONFIG` and `PYTHONHASHSEED` environmental variables in your modifications is concise and focused. 

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Feel free to share additional context or comments if needed. An Ultralytics engineer will review your PR and provide feedback soon. Thank you for your contribution and for helping make Ultralytics even better! 🌟✨

**Comment by Y-T-G** [comment]
Seems like the warnings still appear for tests.

The variable needs to be set before any forward pass. Otherwise, even if it's set, it doesn't get rid of the warning.

**Comment by Y-T-G** [comment]
Moving it inside model init seems to work

**Comment by Y-T-G** [comment]
I ran some tests.

**Effect of setting os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8" on 1 epoch of training on VOC**

| Setting | Deterministic | Warnings | Preprocess Speed | Inference Speed | Loss Speed | Postprocess Speed | CPU Time (User) | CPU Time (Sys) | Total CPU Time | Wall Time |
|---------|----------------|----------|------------------|-----------------|------------|-------------------|-----------------|----------------|----------------|-----------|
| main| True | No | 0.1ms | 0.4ms | 0.0ms | 0.7ms | 2min 52s | 37.1s | 3min 30s | 2min 53s |
| main | False | No | 0.1ms | 0.4ms | 0.0ms | 0.6ms | 2min 45s | 38.3s | 3min 23s | 2min 46s |
| PR  | True | No | 0.1ms | 0.4ms | 0.0ms | 0.7ms | 2min 54s | 39.0s | 3min 33s | 2min 55s |
| PR | False | No | 0.1ms | 0.4ms | 0.0ms | 0.6ms | 2min 42s | 37.1s | 3min 19s | 2min 43s |
| Never set | False | Yes | 0.1ms | 0.4ms | 0.0ms | 0.6ms | 2min 40s | 36.6s | 3min 17s | 2min 41s |

The fastest is not to set it entirely. But this results in warnings during test or deterministic training because it has to be set before any forward pass. Setting it after a forward pass doesn't get rid of the warnings.

The PR sets it, but removes it during non-deterministic training. It's faster than main, but not as fast as without setting it. Although the difference is 2s which could be within margin of error.

**Comment by glenn-jocher** [comment]
This looks good, nice work on the profiling!

**Comment by UltralyticsAssistant** [comment]
🎉 Huge thanks to @Y-T-G for this incredible contribution and to @glenn-jocher for the collaboration! This PR is a testament to how small, thoughtful improvements can create ripples of positive change. 🌟  

As Henry Ford once said, *"Coming together is a beginning, staying together is progress, and working together is success."* 🛠️ Your work improving deterministic training flexibility, reducing warnings, and refining documentation exemplifies that spirit of collaborative success.  

This merge not only enhances flexibility and clarity for our users but also strengthens the foundation we’re building together. 🚀 Your efforts are invaluable—keep up the amazing work! 🙌
